controller/konnect/konnectextension_controller_utils.go

```diff
@@ -33,39 +29,95 @@ import (
 	konnectv1alpha1 "github.com/kong/kubernetes-configuration/api/konnect/v1alpha1"
 )
 
-// getKonnectControlPlane retrieves the Konnect Control Plane based on the provided KonnectExtension specification.
+// getGatewayKonnectControlPlane retrieves the Konnect Control Plane form K8s cluster
+// based on the provided KonnectExtension specification.
 // It supports two types of ControlPlaneRef: KonnectNamespacedRef and KonnectID.
 //
 // Returns:
 // - cp: The retrieved Konnect Control Plane.
 // - res: The result of the controller reconciliation.
 // - err: An error if the retrieval fails.
-func (r *KonnectExtensionReconciler) getKonnectControlPlane(
+func (r *KonnectExtensionReconciler) getGatewayKonnectControlPlane(
 	ctx context.Context,
-	logger logr.Logger,
-	sdk sdkops.ControlPlaneSDK,
 	ext konnectv1alpha1.KonnectExtension,
 	dependingConditions ...metav1.Condition,
-) (cp *sdkkonnectcomp.ControlPlane, res ctrl.Result, err error) {
+) (cp *konnectv1alpha1.KonnectGatewayControlPlane, res ctrl.Result, err error) {
+	// Get respective KonnectGatewayControlPlane from K8s cluster.
+	var getFromK8sErr error
+	switch ext.Spec.Konnect.ControlPlane.Ref.Type {
+	case commonv1alpha1.ControlPlaneRefKonnectNamespacedRef:
+		// TODO: get namespace from cpRef.Namespace when allowed to reference CP from another namespace.
+		cpNN := client.ObjectKey{
+			Name:      ext.Spec.Konnect.ControlPlane.Ref.KonnectNamespacedRef.Name,
+			Namespace: ext.Namespace,
+		}
+		kgcp := &konnectv1alpha1.KonnectGatewayControlPlane{}
+		// Set the controlPlaneRefValidCond to false in case the KonnectGatewayControlPlane is not found.
+		if err := r.Get(ctx, cpNN, kgcp); err != nil {
+			if k8serrors.IsNotFound(err) {
+				getFromK8sErr = err
+			} else {
+				return nil, ctrl.Result{}, err
+			}
+		}
+		cp = kgcp
+	case commonv1alpha1.ControlPlaneRefKonnectID:
+		kgcpList := &konnectv1alpha1.KonnectGatewayControlPlaneList{}
+		if err := r.List(ctx, kgcpList, client.InNamespace(ext.Namespace)); err != nil {
+			return nil, ctrl.Result{}, err
+		}
+		kgcps := lo.Filter(kgcpList.Items, func(kgcp konnectv1alpha1.KonnectGatewayControlPlane, _ int) bool {
+			return kgcp.Status.ID == *ext.Spec.Konnect.ControlPlane.Ref.KonnectID
+		})
```

Comment:
I believe we can use the `IndexFieldKonnectGatewayControlPlaneOnKonnectID` index here to make it a little more efficient, right?